subprojects/ide/src/crossVersionTest/groovy/org/gradle/plugins/ide/tooling/r47/ToolingApiIdeaModelCrossVersionSpec.groovy

```diff
@@ -0,0 +1,203 @@
+        IdeaModule module = project.children[0]
+        IdeaContentRoot root = module.contentRoots[0]
+
+        then:
+        root.sourceDirectories.size() == 1
+        root.sourceDirectories.first().directory == file('src/main/java')
+        root.resourceDirectories.size() == 1
+        root.resourceDirectories.first().directory == file('src/main/resources')
+
+        root.testDirectories.size() == 1
+        root.testDirectories.first().directory == file('src/test/java')
+        root.testResourceDirectories.size() == 1
+        root.testResourceDirectories.first().directory == file('src/test/resources')
+    }
+
+
+    @ToolingApiVersion(">=4.7")
+    @TargetGradleVersion(">=4.7")
+    def "has empty source dirs for project without java plugin"() {
+
+        projectDir.create {
+            src {
+                main {
+                    java {}
+                    resources {}
+                }
+                test {
+                    java {}
+                    resources {}
+                }
+            }
+        }
+
+        when:
+        IdeaProject project = withConnection { connection -> connection.getModel(IdeaProject.class) }
+        IdeaModule module = project.children[0]
+        IdeaContentRoot root = module.contentRoots[0]
+
+        then:
+        root.sourceDirectories.empty
+        root.resourceDirectories.empty
+        root.testDirectories.empty
+        root.testResourceDirectories.empty
+    }
+
+    @ToolingApiVersion(">=4.7")
+    @TargetGradleVersion(">=4.7")
+    def "custom source sets are not added as source directories by default"() {
+
+        buildFile.text = '''
+apply plugin: 'java'
+
+sourceSets {
+    main {
+        java { srcDirs = ['mainSources'] }
+        resources { srcDirs = ['mainResources'] }
+    }
+
+    foo {
+        java { srcDirs = ['fooSources'] }
+        resources { srcDirs = ['fooResources'] }
+    }
+
+    test {
+        java { srcDirs = ['testSources'] }
+        resources { srcDirs = ['testResources'] }
+    }
+
+}
+'''
+
+        projectDir.create {
+            mainSources {}
+            mainResources {}
+            fooSources {}
+            fooResources {}
+            testSources {}
+            testResources {}
+        }
+
+        when:
+        IdeaProject project = withConnection { connection -> connection.getModel(IdeaProject.class) }
+        IdeaModule module = project.children[0]
+        IdeaContentRoot root = module.contentRoots[0]
+
+        then:
+        root.sourceDirectories.size() == 1
+        root.sourceDirectories[0].directory == file('mainSources')
+        root.resourceDirectories.size() == 1
+        root.resourceDirectories[0].directory == file('mainResources')
+        root.testDirectories.size() == 1
+        root.testDirectories[0].directory == file('testSources')
+        root.testResourceDirectories.size() == 1
+        root.testResourceDirectories[0].directory == file('testResources')
+    }
+
```

Comment:
Add test coverage to check that all new methods throw `UnsupportedOperationException` when older Gradle distributions are used.

```
@TargetGradleVersion("<4.7")
...
when:
root.getResourceDirectories()

then:
thrown UnsupportedOperationException
```